lib/maths/CMic.cc

```diff
@@ -0,0 +1,262 @@
+    // information given in "An Empirical Study of Leading Measures of Dependence"
+    // Reshef et al. See table E-4.
+
+    double n{std::max(std::min(static_cast<double>(m_Samples.size()), 5000.0), 50.0)};
+
+    double x[]{49.0, 50.0, 100.0, 500.0, 1000.0, 5000.0, 5001.0};
+    double f[]{5.0, 5.0, 5.0, 5.0, 4.0, 1.0, 1.0};
+
+    auto b = std::lower_bound(std::begin(x), std::end(x), n);
+    auto a = b - 1;
+
+    double fb{f[b - std::begin(x)]};
+    double fa{f[a - std::begin(x)]};
+
+    return CTools::linearlyInterpolate(*a, *b, fa, fb, n);
+}
+
+CMic::TDoubleVec CMic::equipartitionAxis(std::size_t variable, std::size_t l) const {
+
+    // We seek an l-partition of the y-axis into buckets of approximately equal
+    // numbers of points.
+
+    LOG_TRACE(<< "Equipartition " << (variable == 0 ? "x" : "y") << "-axis");
+
+    TDoubleVec result;
+    result.reserve(l + 1);
+
+    std::size_t desired{std::max((m_Samples.size() + l - 1) / l, std::size_t{1})};
+    LOG_TRACE(<< "l = " << l << ", desired bucket size = " << desired);
+
+    for (std::size_t i = desired; i < m_Order[variable].size(); i += desired) {
+        result.push_back(m_Samples[m_Order[variable][i]](variable));
+    }
+    result.erase(std::unique(result.begin(), result.end()), result.end());
+    double max{m_Samples[m_Order[variable].back()](variable)};
+    result.push_back((1.0 + std::copysign(EPS, max)) * max);
+    LOG_TRACE(<< "partition = " << core::CContainerPrinter::print(result));
+
+    return result;
+}
+
+CMic::TDoubleVec CMic::optimizeXAxis(const TDoubleVec& q, std::size_t l, std::size_t k) const {
+
+    // Compute the k-partition of the x-axis which maximises the mutual information
+    // between the variables given a fixed l-partition of the y-axis.
+
+    using TDoubleVecVec = std::vector<TDoubleVec>;
+
+    LOG_TRACE(<< "Optimize x-axis, k = " << k);
+
+    std::size_t ck{static_cast<std::size_t>(this->c() * k)};
+    LOG_TRACE(<< "c * k = " << ck);
+
+    TDoubleVec pi(this->equipartitionAxis(X, ck));
+    LOG_TRACE(<< "pi = " << core::CContainerPrinter::print(pi));
+    ck = std::min(ck, pi.size());
+
+    double n{static_cast<double>(m_Samples.size())};
+    double w{1.0 / n};
+
+    // Compute grid cell and row and column probabilities.
+    TVectorXdVec cump(ck, TVectorXd{l});
+    for (const auto& sample : m_Samples) {
+        std::ptrdiff_t i{std::upper_bound(pi.begin(), pi.end(), sample(X)) - pi.begin()};
+        std::ptrdiff_t j{std::upper_bound(q.begin(), q.end(), sample(Y)) - q.begin()};
+        cump[i](j) += w;
+    }
+
+    // We need to cumulative probabilities in each column for the dynamic program.
+    for (std::size_t i = 1; i < ck; ++i) {
+        cump[i] += cump[i - 1];
+    }
+
+    TDoubleVecVec M(ck);
+
+    // Compute the optimal 2-subset of the x-axis partition pi.
+    for (std::size_t t = 2; t <= ck; ++t) {
+
+        double Fmax{-INF};
+
+        double Z{cump[t - 1].L1()};
```

Comment:
Good point. It'll be interesting how this stacks up since this will be vectorised and the vectors are small it may be relatively few cycles to just recompute, but it does look like caching should be a win. I'll try it out.

Reply:
This worked out well, saving a little over 5% of the runtime of unit tests for example, see [e9a441e](https://github.com/elastic/ml-cpp/pull/488/commits/e9a441eec06cb752368b0cc570887bdc339b630a).